File: apps/api/app/api/schemas/moderation.py
"""Pydantic schemas for Moderation API."""

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.models import EntityType, TaskStatus


class ModerationTaskResponse(BaseModel):
    """Response schema for moderation task."""

    id: uuid.UUID
    entity_type: EntityType
    entity_id: uuid.UUID
    org_id: uuid.UUID | None = None
    status: TaskStatus
    priority: int = 0
    reason: str | None = None
    created_by_user_id: uuid.UUID | None = None
    moderator_user_id: uuid.UUID | None = None
    created_at: datetime
    decided_at: datetime | None = None

    model_config = {"from_attributes": True}


class ModerationTaskDetailResponse(BaseModel):
    """Response schema for moderation task with entity details."""

    task: ModerationTaskResponse
    entity_summary: dict[str, Any]


class ModerationTaskListResponse(BaseModel):
    """Response schema for list of moderation tasks."""

    items: list[ModerationTaskResponse]
    total: int
    limit: int
    offset: int


class ModerationTaskListParams(BaseModel):
    """Query parameters for listing moderation tasks."""

    entity_type: EntityType | None = None
    status: TaskStatus | None = None
    org_id: uuid.UUID | None = None
    limit: int = Field(default=50, ge=1, le=100)
    offset: int = Field(default=0, ge=0)


class RejectRequest(BaseModel):
    """Request schema for rejecting a moderation task."""

    reason: str = Field(..., min_length=1, max_length=1000)


class ApproveResponse(BaseModel):
    """Response schema for approve action."""

    message: str
    task: ModerationTaskResponse


class RejectResponse(BaseModel):
    """Response schema for reject action."""

    message: str
    task: ModerationTaskResponse

File: apps/web/src/lib/api/moderation.ts


import { apiClient } from './client';


export type EntityType = 'page' | 'media';
export type TaskStatus = 'pending' | 'approved' | 'rejected';

export interface ModerationTaskDTO {
  id: string;
  entity_type: EntityType;
  entity_id: string;
  org_id: string | null;
  status: TaskStatus;
  priority: number;
  reason: string | null;
  created_by_user_id: string | null;
  moderator_user_id: string | null;
  created_at: string;
  decided_at: string | null;
}

export interface EntitySummary {
  entity_type: string;
  entity_id: string;
  name?: string;
  slug?: string;
  title?: string;
  status?: string;
  visibility?: string;
  biography_preview?: string;
  media_type?: string;
  filename?: string;
  url?: string;
  moderation_status?: string;
  object_type?: string;
  description_preview?: string;
  lat?: number;
  lng?: number;
  address?: string;
  page_id?: string;
  created_at?: string;
}

export interface ModerationTaskDetailDTO {
  task: ModerationTaskDTO;
  entity_summary: EntitySummary;
}

export interface ModerationTaskListResponse {
  items: ModerationTaskDTO[];
  total: number;
  limit: number;
  offset: number;
}

export interface ListTasksParams {
  entity_type?: EntityType | undefined;
  status?: TaskStatus | undefined;
  org_id?: string | undefined;
  limit?: number | undefined;
  offset?: number | undefined;
}

export interface RejectPayload {
  reason: string;
}

export interface ActionResponse {
  message: string;
  task: ModerationTaskDTO;
}


export const adminModerationApi = {
  listTasks: async (params?: ListTasksParams): Promise<ModerationTaskListResponse> => {
    const query = new URLSearchParams();
    if (params?.entity_type) query.set('entity_type', params.entity_type);
    if (params?.status) query.set('status', params.status);
    if (params?.org_id) query.set('org_id', params.org_id);
    if (params?.limit) query.set('limit', String(params.limit));
    if (params?.offset) query.set('offset', String(params.offset));
    const qs = query.toString();
    return apiClient.get<ModerationTaskListResponse>(`/api/admin/moderation/tasks${qs ? `?${qs}` : ''}`);
  },

  getTask: async (taskId: string): Promise<ModerationTaskDetailDTO> => {
    return apiClient.get<ModerationTaskDetailDTO>(`/api/admin/moderation/tasks/${taskId}`);
  },

  approveTask: async (taskId: string): Promise<ActionResponse> => {
    return apiClient.post<ActionResponse>(`/api/admin/moderation/tasks/${taskId}/approve`);
  },

  rejectTask: async (taskId: string, data: RejectPayload): Promise<ActionResponse> => {
    return apiClient.post<ActionResponse>(`/api/admin/moderation/tasks/${taskId}/reject`, data);
  },
};


export const orgModerationApi = {
  listTasks: async (orgId: string, params?: Omit<ListTasksParams, 'org_id'>): Promise<ModerationTaskListResponse> => {
    const query = new URLSearchParams();
    if (params?.entity_type) query.set('entity_type', params.entity_type);
    if (params?.status) query.set('status', params.status);
    if (params?.limit) query.set('limit', String(params.limit));
    if (params?.offset) query.set('offset', String(params.offset));
    const qs = query.toString();
    return apiClient.get<ModerationTaskListResponse>(`/api/orgs/${orgId}/moderation/tasks${qs ? `?${qs}` : ''}`);
  },

  getTask: async (orgId: string, taskId: string): Promise<ModerationTaskDetailDTO> => {
    return apiClient.get<ModerationTaskDetailDTO>(`/api/orgs/${orgId}/moderation/tasks/${taskId}`);
  },

  approveTask: async (orgId: string, taskId: string): Promise<ActionResponse> => {
    return apiClient.post<ActionResponse>(`/api/orgs/${orgId}/moderation/tasks/${taskId}/approve`);
  },

  rejectTask: async (orgId: string, taskId: string, data: RejectPayload): Promise<ActionResponse> => {
    return apiClient.post<ActionResponse>(`/api/orgs/${orgId}/moderation/tasks/${taskId}/reject`, data);
  },
};
